# Lab | Hypothesis Testing

**Objective**

Welcome to the Hypothesis Testing Lab, where we embark on an enlightening journey through the realm of statistical decision-making! In this laboratory, we delve into various scenarios, applying the powerful tools of hypothesis testing to scrutinize and interpret data.

From testing the mean of a single sample (One Sample T-Test), to investigating differences between independent groups (Two Sample T-Test), and exploring relationships within dependent samples (Paired Sample T-Test), our exploration knows no bounds. Furthermore, we'll venture into the realm of Analysis of Variance (ANOVA), unraveling the complexities of comparing means across multiple groups.

So, grab your statistical tools, prepare your hypotheses, and let's embark on this fascinating journey of exploration and discovery in the world of hypothesis testing!

**Challenge 1**

In this challenge, we will be working with pokemon data. The data can be found here:

- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/pokemon.csv

In [1]:
#libraries
import pandas as pd
import scipy.stats as st
import numpy as np


In [2]:
df = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/pokemon.csv")
df

,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
0,Bulbasaur,Grass,Poison,45,49,49,65,65,45,1,False
1,Ivysaur,Grass,Poison,60,62,63,80,80,60,1,False
2,Venusaur,Grass,Poison,80,82,83,100,100,80,1,False
3,Mega Venusaur,Grass,Poison,80,100,123,122,120,80,1,False
4,Charmander,Fire,NaN,39,52,43,60,50,65,1,False
...,...,...,...,...,...,...,...,...,...,...,...
795,Diancie,Rock,Fairy,50,100,150,100,150,50,6,True
796,Mega Diancie,Rock,Fairy,50,160,110,160,110,110,6,True
797,Hoopa Confined,Psychic,Ghost,80,110,60,150,130,70,6,True
798,Hoopa Unbound,Psychic,Dark,80,160,60,170,130,80,6,True


- We posit that Pokemons of type Dragon have, on average, more HP stats than Grass. Choose the propper test and, with 5% significance, comment your findings.

In [3]:
# H0: mean(HP | Dragon) <= mean(HP | Grass)
# H1: mean(HP | Dragon) >  mean(HP | Grass)
# Two independent samples, unknown/unequal variances -> Welch's t-test, one-tailed

dragon_hp = df[df['Type 1'] == 'Dragon']['HP']
grass_hp = df[df['Type 1'] == 'Grass']['HP']

t_stat, p_two_tailed = st.ttest_ind(dragon_hp, grass_hp, equal_var=False)
# one-tailed p-value: halve the two-tailed p, only valid because t_stat > 0
# (i.e. the sample already points in the direction of H1)
p_one_tailed = p_two_tailed / 2 if t_stat > 0 else 1 - p_two_tailed / 2

print(f"Dragon: n={len(dragon_hp)}, mean HP={dragon_hp.mean():.2f}")
print(f"Grass:  n={len(grass_hp)}, mean HP={grass_hp.mean():.2f}")
print(f"t-statistic: {t_stat:.4f}")
print(f"one-tailed p-value: {p_one_tailed:.6f}")
print(f"Reject H0 at 5%? {p_one_tailed < 0.05}")

Dragon: n=32, mean HP=83.31
Grass:  n=70, mean HP=67.27
t-statistic: 3.3350
one-tailed p-value: 0.000799
Reject H0 at 5%? True


**Finding:** Dragon-type Pokémon average 83.3 HP vs. 67.3 HP for Grass-type — a real gap, and the one-tailed Welch's t-test confirms it's not noise (p ≈ 0.0008, well under 0.05). We reject H0: Dragon-type Pokémon do have significantly higher HP than Grass-type on average.

- We posit that Legendary Pokemons have different stats (HP, Attack, Defense, Sp.Atk, Sp.Def, Speed) when comparing with Non-Legendary. Choose the propper test and, with 5% significance, comment your findings.


In [4]:
# H0: mean(stat | Legendary) == mean(stat | Non-Legendary), for each stat
# H1: mean(stat | Legendary) != mean(stat | Non-Legendary)
# Only 2 groups (Legendary vs not) but 6 separate outcome variables -> one
# two-sided Welch's t-test per stat, not ANOVA (ANOVA is for 3+ groups on
# a single outcome).

stats_cols = ['HP', 'Attack', 'Defense', 'Sp. Atk', 'Sp. Def', 'Speed']
legendary = df[df['Legendary']]
non_legendary = df[~df['Legendary']]

results = []
for col in stats_cols:
    t_stat, p_val = st.ttest_ind(legendary[col], non_legendary[col], equal_var=False)
    results.append({
        'stat': col,
        'legendary_mean': legendary[col].mean(),
        'non_legendary_mean': non_legendary[col].mean(),
        't_stat': t_stat,
        'p_value': p_val,
        'significant_5pct': p_val < 0.05,
    })

results_df = pd.DataFrame(results)
results_df

,stat,legendary_mean,non_legendary_mean,t_stat,p_value,significant_5pct
0,HP,92.738462,67.182313,8.981370,1.002691e-13,True
1,Attack,116.676923,75.669388,10.438134,2.520372e-16,True
2,Defense,99.661538,71.559184,7.637078,4.826998e-11,True
3,Sp. Atk,122.184615,68.454422,13.417450,1.551461e-21,True
4,Sp. Def,105.938462,68.892517,10.015697,2.294933e-15,True
5,Speed,100.184615,65.455782,11.475044,1.049016e-18,True


**Finding:** All six stats come back significant at 5% (every p-value is far below 0.05, several under 1e-10) — Legendary Pokémon are stronger across the board, not just on one or two stats. `Sp. Atk` shows the largest gap in significance (p ≈ 1.6e-21), `Defense` the smallest of the six (p ≈ 4.8e-11), but even that is overwhelming evidence against H0.

**Challenge 2**

In this challenge, we will be working with california-housing data. The data can be found here:
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/california_housing.csv

In [5]:
df2 = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/california_housing.csv")
df2.head()

,longitude,latitude,housing_median_age,total_rooms,total_bedrooms,population,households,median_income,median_house_value
0,-114.31,34.19,15.0,5612.0,1283.0,1015.0,472.0,1.4936,66900.0
1,-114.47,34.40,19.0,7650.0,1901.0,1129.0,463.0,1.8200,80100.0
2,-114.56,33.69,17.0,720.0,174.0,333.0,117.0,1.6509,85700.0
3,-114.57,33.64,14.0,1501.0,337.0,515.0,226.0,3.1917,73400.0
4,-114.57,33.57,20.0,1454.0,326.0,624.0,262.0,1.9250,65500.0


**We posit that houses close to either a school or a hospital are more expensive.**

- School coordinates (-118, 34)
- Hospital coordinates (-122, 37)

We consider a house (neighborhood) to be close to a school or hospital if the distance is lower than 0.50.

Hint:
- Write a function to calculate euclidean distance from each house (neighborhood) to the school and to the hospital.
- Divide your dataset into houses close and far from either a hospital or school.
- Choose the propper test and, with 5% significance, comment your findings.
 

In [6]:
def euclidean_distance(lon, lat, ref_lon, ref_lat):
    return np.sqrt((lon - ref_lon) ** 2 + (lat - ref_lat) ** 2)

SCHOOL = (-118, 34)
HOSPITAL = (-122, 37)

df2['dist_school'] = euclidean_distance(df2['longitude'], df2['latitude'], *SCHOOL)
df2['dist_hospital'] = euclidean_distance(df2['longitude'], df2['latitude'], *HOSPITAL)

df2['close'] = (df2['dist_school'] < 0.50) | (df2['dist_hospital'] < 0.50)

close_value = df2[df2['close']]['median_house_value']
far_value = df2[~df2['close']]['median_house_value']

print(f"Close: n={len(close_value)}, mean value=${close_value.mean():,.0f}")
print(f"Far:   n={len(far_value)}, mean value=${far_value.mean():,.0f}")

Close: n=6829, mean value=$246,952
Far:   n=10171, mean value=$180,678


In [7]:
# H0: mean(value | close) <= mean(value | far)
# H1: mean(value | close) >  mean(value | far)
# Two independent samples, unequal variances -> Welch's t-test, one-tailed

t_stat, p_two_tailed = st.ttest_ind(close_value, far_value, equal_var=False)
p_one_tailed = p_two_tailed / 2 if t_stat > 0 else 1 - p_two_tailed / 2

print(f"t-statistic: {t_stat:.4f}")
print(f"one-tailed p-value: {p_one_tailed:.3e}")
print(f"Reject H0 at 5%? {p_one_tailed < 0.05}")

t-statistic: 37.9923
one-tailed p-value: 1.503e-301
Reject H0 at 5%? True


**Finding:** Houses close to a school or hospital average **$246,952**, vs. **$180,678** for houses farther away — about 37% more expensive. The one-tailed Welch's t-test rejects H0 overwhelmingly (p ≈ 1.5e-301, nowhere near the 5% threshold). Proximity to a school or hospital is strongly associated with higher median house value in this dataset — though as with any observational data, this is an association, not proof that proximity itself *causes* the price premium (it could just as easily reflect that schools/hospitals tend to sit in already-desirable neighborhoods).